src/include/platform/internal/GenericConfigurationManagerImpl.ipp

```diff
@@ -586,6 +586,57 @@ CHIP_ERROR GenericConfigurationManagerImpl<ConfigClass>::SetFailSafeArmed(bool v
     return WriteConfigValue(ConfigClass::kConfigKey_FailSafeArmed, val);
 }
 
+template <class ConfigClass>
+CHIP_ERROR GenericConfigurationManagerImpl<ConfigClass>::GetDeviceLocation(DeviceLocatioType & location)
+{
+    uint8_t locationData[kMaxDeviceLocationNameLength + sizeof(DeviceLocatioType)];
+    MutableByteSpan locationSpan(locationData);
+
+    size_t outLen = 0;
+    ReturnErrorOnFailure(
+        ReadConfigValueBin(ConfigClass::kConfigKey_DeviceLocation, locationSpan.data(), locationSpan.size(), outLen));
+
+    TLV::TLVReader tlvReader;
+    tlvReader.Init(locationSpan);
+
+    char locationNameBuffer[DeviceLayer::ConfigurationManager::kMaxDeviceLocationNameLength] = { 0 };
+    DeviceLocatioType loc({
+        .locationName = MutableCharSpan(locationNameBuffer),
+    });
+
+    ReturnErrorOnFailure(tlvReader.Next(TLV::AnonymousTag()));
+    ReturnErrorOnFailure(loc->Decode(tlvReader));
+
+    if (loc.IsNull())
+    {
+        location.SetNull();
+        return CHIP_NO_ERROR;
+    }
+
+    memcpy((void *) location.Value().locationName.data(), loc.Value().locationName.data(), loc.Value().locationName.size());
```

Comment:
@bzbarsky-apple Is there a simpler way of achieving this? I was expecting the `Decode` method to copy the contents of the `tlvReader` to the `location` argument. Maybe there is a different `Decode` method that does this?

Reply:
Here's the thing.  `DeviceLocatioType` [sic] is an alias for `DeviceLocation::TypeInfo::Type`, which is `Nullable<Structs::HomeLocationStruct::Type>`.  And that's defined as:

```
struct Type
{
public:
    chip::CharSpan locationName;
```

so `location.Value().locationName` is a `const char *`.

The code above is hiding that "oh, I am copying to a pointer that points to const data, which is generally not OK" with that C-style cast (which I bet got added because the compiler complained about passing a const pointer to a non-const `void*` arg!), but there is absolutely no way to tell from the code in this function that this memcpy business is safe, much less telling it from inside the `Decode` method or anything like that.

The normal way I would approach this is:

1. Define a struct type inheriting from `DeviceLocatioType` [sic] which has storage for the data.
2. Define a copy assignment operator on that which takes `DeviceLocatioType` [sic] and copies things correctly.
3. Pass in a reference to this new type.
4. Decode into a `DeviceLocatioType` on the stack and then assign into the outparam.